### ISSUE TITLE
Cluttered ways

### ISSUE THREAD

**Issue by vyskocil** [Framstag/libosmscout#1220]
Way rendering maybe cluttered, especially the one way as you could see in the attached screenshot the way labels and one way arrows often overlaps. Is there a solution to avoid this issue ? 
Maybe the way rendering code should treat the labels and the images in the same process allowing to alternate both along the path ? 
Another simpler solution could be to provide the background color of the way to the label renderer, thus it could use it to draw the text with opaque background and overdraw the one way arrows.

![Cluttered ways](https://user-images.githubusercontent.com/452704/155505079-9a2dc839-c710-4c1d-b572-b73ce97c0284.png)

**Comment by vyskocil** [comment]
Here is the standard OpenStreetMap rendering of the same area : 
 
<img width="752" alt="Screenshot 2022-02-24 at 11 23 46" src="https://user-images.githubusercontent.com/452704/155506107-b3618353-f771-4511-a7e1-2ea836122bf5.png">

**Comment by Framstag** [comment]
This (coordinated rendering of symbols ad labels) should in principle possible, however would need some changes.

Currently labels are handled by DrawWayContourLabel, with positioning done in the LabelLayouter.RegisterContourLabel method, based on parameter passed in MapPainter::DrawWayContourLabel.

For symbols this is triggered in MapPainter::DrawWayDecoration with the actual positioning in MapPainterXXX.DrawContourLabel()

In both cases, start and offset to the next instance just would to be input from the outside (and calculated for both coordinated), instead being calculated separately.

**Comment by Framstag** [comment]
@vyskocil 

The placement for the contour symbols has now been moved (in principle, did not check all backends - only cairo and Qt) from the individual backend code to the MapPainter.
Similar refactoring would have to be done now for the label placement of the way labels.
Finally, this would allow coordinating the placement of labels and contour symbols.

Note, that contour symbols now can be scaled in height to the width of the way.

If some of my refactoring broke the IOSX backend, please contact me. This should be easy to fix, just make sure to evaluate alle the information in SymbolContourData.

**Comment by vyskocil** [comment]
Hi @Framstag ! Thank you very much taking this issue into account ! 
I just tried the up to date code just before the scalable contour symbol which is not merged yet, there is indeed a lot less clutter as in my case there is no more way labels at all... ;-) Maybe there are some parameters to tweak in the stylesheet file (I don't use the one provided with the code) ?

Here is a screenshot of the same place : 

![IMG_EBE6AB3F66D5-1](https://user-images.githubusercontent.com/452704/190896936-ee8e0f52-3cf7-459d-a823-8dc4bc97f07d.jpeg)

**Comment by vyskocil** [comment]
Looking further on the map rendered on iOS I couldn't find any way label now, even for the no one way roads... What should I look for to debug this issue ?

**Comment by Framstag** [comment]
@vyskocil AFAIK I did not do any principle changes to the text handling. I had to change/enhance a few APIs regarding Symbols but that should not have had any influence on the text APIs.

Since I normally had to "poke" the source for IOS without Syntax higlighting, local build or testing, it is possible that I broke something. I suggest you check my change around the LabelLayouter to check, if I did a wrong replacement that compiled but broke something.

Since I do not regularly check to correctness of all the backends, it is possible, though, that it broke before.

Sorry for that possible inconvenience.

Yes, still sorting out build errors, may take some time into the week to get that fixed...time is already running out :-/

**Comment by vyskocil** [comment]
@Framstag  regarding the scalable contour symbol, I'm trying to add the needed code to support this option for iOS/OSX backend but I'm having an issue. 
I modified void MapPainterIOS::DrawContourSymbol() to pass the new scaleFactor argument to DrawSymbol : 

```
...
DrawSymbol(projection, parameter, symbol, 0, 0, data.symbolScale);
...
```

But in DrawSymbol code I only receive 1.0 or 0.0 values for scaleFactor !

I also modified the stylesheet file with : 

```
   WAY.SYMBOL { symbol: oneway_arrow; symbolSpace: 10mm; renderMode: scale; scale: 0.5;}
```

**Comment by Framstag** [comment]
@vyskocil 

The necessary algorithm is implemented in MapPainter::DrawWayDecoration (else branch => more refactoring needed ;-)):

```
        if (pathSymbolStyle->GetRenderMode()==PathSymbolStyle::RenderMode::scale) {
          symbolScale=data.mainSlotWidth*pathSymbolStyle->GetScale()/symbolBoundingBox.GetHeight();
        }
        else {
          symbolScale=1.0;
        }
```
The stylesheet entry should result in the if-branch getting entered.

It sets symbolScale and then calls DrawContourSymbol(). DrawContourSymbol itself is driver specific and must handle the scale factor. You can implement SymbolRenderer and delegate rendering to it, but you do not need to (see the cairo version to see, how I integrated the transformation code).

Hope this helps.

**Comment by vyskocil** [comment]
@Framstag does that mean that I have to wait the other branch is refactored and merged to receive the good symbolScale in the iOS driver DrawContourSymbol method ?

**Comment by Framstag** [comment]
No, the currently last not merged branch only contains activation of scaled contour symbols for the Cairo backend. It does though contain some refactoring o the SymbolRenderer.

I'm just waiting for the build to finish...to merge it.

**Comment by vyskocil** [comment]
Ok thanks, I'll try it again when all is merged but as I understand it, it should not change much. 
Actually I'm not using the SymbolRenderer in the iOS/OSX backend, I only updated the previous DrawSymbol code in the iOS driver adding code to append a affine transform before drawing to scale the symbol accordingly but if it receive 0 it would not work :-) 
Then I think there is something not working when osmscout-map call the iOSX driver maybe there is a issue with : 
```
symbolScale=data.mainSlotWidth*pathSymbolStyle->GetScale()/symbolBoundingBox.GetHeight();
```
Maybe this code use some wrong value from the iOS driver ?

**Comment by Framstag** [comment]
Hmm, symbolBoundingBox.GetHeight() is general symbol code. pathSymbolStyle->GetScale() comes straight from the stylesheet. mainSlotWidth is calculated in CalculateWayPaths() and should not have driver-specific parameters.

Attached an image to show how it would look like...

![test](https://user-images.githubusercontent.com/14061582/190916957-34c1937c-519b-4df5-a70d-d2f9e0871eab.png)

**Comment by Framstag** [comment]
This is https://www.openstreetmap.org/#map=18/51.51632/7.46974

**Comment by Framstag** [comment]
Merged

**Comment by vyskocil** [comment]
Hi @Framstag ! 
I put some debug logs in MapPainter.cpp and here's what I'm getting for the symbolScale calculation :

```symbolScale=0  data.mainSlotWidth=0 pathSymbolStyle->GetScale()=0.5 symbolBoundingBox.GetHeight()=21.5433```

data.mainSlotWidth is always 0 in these cases...

**Comment by Framstag** [comment]
mainSlotWidth gets calculated in CalculateWayPaths(). There it iterates through the line styles. The line style with the empty slot is the "main" style. It defines the width of the way, while other, additional line styles (!slot.empty()) are decorating and likely also have an offset to the center of the way (though all this depends on the actual style definition).

Also StyleConfig::GetWayLineStyles() should do some sorting to assure that the style with the empty slot is the first in the list.

It is also important, that further calculations are done based on the mainSlotWidth further down CalculateWayPaths().

The code looks a little fragile because it seems it only works correctly under the assumption that mainSlotWidth is calculated based on the first line style in the list and that this line style has an empty slot. If this is not the case, things can possibly partly break. This should be easy to check with some additional checking code. 

Or CalculateLineWith() always returns 0.0 because of strange style sheet (no width given) or strange effects with the projection class.

If you have a style sheet, that is different from the default one, you can send it to me per mail, and I'll check, if it works for me.  

It would be good, if we would have a DrawMapIOSX demo application. This way, it would be simple for me to reproduce and analyze the problem, since I still have a Mac Book.

I hope this help you to further debug the information. If you have an idea, how I can easily reproduce the problem on the Mac (without any in deep knowledge of IOSX development and the Apple development environment), give me a hint. I'LL to take a look at it in this case, too.

**Comment by vyskocil** [comment]
Hi @Framstag ! 

I tried the iOS renderer using the standard stylesheet provided in the distribution (just a bit tweaked to remove some types not imported in the database I have on hands), I don't get data.mainSlotWidth == 0 anymore and could see the one way icons ! But there are still very few, almost none road names displayed, even for non one way roads...
Here is a screenshot:

![IMG_1575C9165ED1-1](https://user-images.githubusercontent.com/452704/192236306-37648f42-8a79-40c9-9293-51bf5398c5bf.jpeg)

**Comment by Framstag** [comment]
So it sounds like there is something triggered, dependent on the actual style sheet. I'll take a look, if I can make the relevant code more robust. We will see if this will help.

**Comment by Framstag** [comment]
Regarding the text label rendering...

Please check, if MapPainter::DrawWayContourLabel() can extract a label. If yes, this should call DrawWayContourLabel() which will create a contour label entry for it using MapPainter::RegisterContourLabel() (which is Renderer specific). It should just call your labelLayouter.

The list of labels as placed by the LabelLayouter is later on rendered by LabelLayouter::DrawLabels(), calling DrawLabel() and/or DrawGlyphs() back on your Renderer.

Please check at the mentioned code places, if there are still street labels to process. Either they get lost somewhere, or font size is too small to be seeen, color is not visible (alpha!) or similar.

**Comment by Framstag** [comment]
@vyskocil : See https://github.com/Framstag/libosmscout/pull/1323. Please check, if this changes anything regarding mainSlotWidth. It should either fix the problem or at least print out a warning (if the logger is configured) with the problematic way and its type. This should allow in turn to check the style sheet for an error.

**Comment by Karry** [comment]
Hi @Framstag , I found several issues with your changes from last week. It is visible even with Qt backend. I prepared simple bash script:
```
#!/bin/bash -x

while [ $# -gt 0 ] ; do
  git checkout "$1"
  suffix="$(git log --pretty=format:'%ci %h' | head -1)"
  make -j 8 DrawMapQt
  ./Demos/DrawMapQt \
    --debug \
    --baseMap /home/karry/Maps/world \
    --dpi 200 \
    ~/Maps/europe-france-20220905-004309 \
    ../stylesheets/standard.oss \
    1920 1080 \
    43.61996 6.969865 \
    524288 \
    "DrawMapQt $suffix.png"
  shift
done
```
And run this script for last 500 commits in master (`git log --pretty=format:'%h' | head -500 | xargs -n 1 ../draw-commit.sh`). 

Starting with ea3625027, result is something I am experienced 
![DrawMapQt 2022-09-05 10:18:29 +0200 ea3625027](https://user-images.githubusercontent.com/309458/192468796-a785756d-e44c-4ebe-b7e3-3431ee866537.png)

Street labels disappear with e3208ea1c
![DrawMapQt 2022-09-17 14:03:13 +0200 e3208ea1c](https://user-images.githubusercontent.com/309458/192469102-f8386209-9d31-4562-9de0-3332ac439b4d.png)

Restaurant symbols disappears at e1deda23e
![DrawMapQt 2022-09-17 14:06:53 +0200 e1deda23e](https://user-images.githubusercontent.com/309458/192472622-2a6a2247-6e04-46cd-bb25-d7f06c05a40e.png)

Church symbol is moved from 91991eb11
![DrawMapQt 2022-09-17 20:42:12 +0200 91991eb11](https://user-images.githubusercontent.com/309458/192475263-2af7f3c4-8cd1-4434-a81c-8fedaad4960b.png)

And finally, oneway arrows are rendered too git with high density from bcbaa660e
![DrawMapQt 2022-09-18 09:59:27 +0200 bcbaa660e](https://user-images.githubusercontent.com/309458/192475507-d18bcf3e-aa40-4278-aef1-995dac573c63.png)

Should I create tickets for individual problems?

**Comment by Framstag** [comment]
Yes, please make tickets. This should help be fix the individual problems.

**Comment by Framstag** [comment]
Given the reproducer infos by @Karry, I was able to reproduce the problem locally. Thanks. I'll take a look.

**Comment by vyskocil** [comment]
When zooming a lot on small roundabout with the standard stylesheet I got this on iPhone : 

![IMG_C5FDA5F37117-1](https://user-images.githubusercontent.com/452704/193095140-7aece4e4-40b8-4ff6-a216-59c4b1d968c0.jpeg)

**Comment by vyskocil** [comment]
> > @vyskocil : See #1323. Please check, if this changes anything regarding mainSlotWidth. It should either fix the problem or at least print out a warning (if the logger is configured) with the problematic way and its type. This should allow in turn to check the style sheet for an error.
> 
> I rebased #1323 on the latest code and tried it and there was no warning

I removed #1323 and now with my own stylesheet the oneway symbol are not drawn anymore but they are still there with the standard stylesheet

**Comment by Framstag** [comment]
So with my path, things work better with the standard stylesheet (overlay arrows appear), though problem is always there with your stylesheet.

Sounds like my fix improves the situation (and I should merge), while your style sheet seems to have a problem?

In this case you should get a warning from the renderer (seee patch):

    if (pathData.mainSlotWidth==0.0) {
      log.Warn() << "Line style for way " << way.GetFileOffset() << " of type " << way.GetType()->GetName() << " results in empty mainSlotWidth";
    }

Are you sure logging is active? Else: Can you set a break point (either on the condition or the log output) to check?

**Comment by Framstag** [comment]
@vyskocil : Which parts of the issue are still relevant?

**Comment by vyskocil** [comment]
I'm sorry but I forgot this issue, I'll look at this again. I tried to display the same area and we could still see overlapping symbols and road names, maybe there is some work to be done in the iOS/OSX backend.

<img width="330" height="717" alt="Image" src="https://github.com/user-attachments/assets/c0834d4e-49d2-4d47-893f-79a51c119de8" />
